### ISSUE TITLE
freeze_bn config value

### ISSUE THREAD

**Issue by emjay73** [dbolya/yolact#228]
In the code, I found some lines that confuse me.

train.py 

    if not cfg.freeze_bn: yolact_net.freeze_bn() # Freeze bn so we don't kill our means
    yolact_net(torch.zeros(1, 3, cfg.max_size, cfg.max_size).cuda())
    if not cfg.freeze_bn: yolact_net.freeze_bn(True)

why is the batch norm frozen when the freeze_bn is actually 'False'.
There is a mismatch between configuration and actual behavior, I guess.

**Comment by dbolya** [comment]
That piece of code is freezing batch norm, initializing the network with a dummy tensor, than unfreezing batchnorm. The point of freezing batch norm there is so that the dummy tensor doesn't affect our batch norm means, but I guess it's sort of a hack way to do that.

**Comment by emjay73** [comment]
Then what is this dummy inference for? 
What kinds of parameters are going to be initialized with eval mode?

**Comment by dbolya** [comment]
The whole network is not on eval mode during initialization, just the batch norm. This will initialize the graph structure, prototypes, and other cached stuff. It wasn't as important when using 1 gpu before since you could just initialize all that in the first iteration, but when I added real support for multiple GPUs it became important because suddenly it was initializing those things every iteration (for each iteration, dataparallel basically clones out your model, does one gradient update, and destroys the cloned versions before the next iteration, so if you don't initialize before using dataparallel, it's going to initialize every iteration, which makes thing very slow).

**Comment by emjay73** [comment]
Well, It's kinda vague for me.
If you see the train.py code, 

----------------------------------------------------------------------------------------------------------------
    net = CustomDataParallel(NetLoss(net, criterion))
    if args.cuda:
        net = net.cuda()
    
    # Initialize everything
    if not cfg.freeze_bn: yolact_net.freeze_bn() # Freeze bn so we don't kill our means
    yolact_net(torch.zeros(1, 3, cfg.max_size, cfg.max_size).cuda())
    if not cfg.freeze_bn: yolact_net.freeze_bn(True)
----------------------------------------------------------------------------------------------------------------

objects called 'net' and yolact_net have different IDs, so I'm not sure what is the effect of yolact_net initialization for the sake of 'net' initialization.

**Comment by dbolya** [comment]
net contains yolact_net as a member. That "net" being passed into the CustomDataParallel is yolact_net.

All I'm doing is forcefully initializing the yolact_net part of net because it has information that it stores inside during initialization.

**Comment by emjay73** [comment]
Ok. Now I get the relationship between 'net' and 'yolact_net'.
And what you are saying is inferencing with zero tensor can initialize yolact_net, and initialized values like graph structure, prototypes(??), and other cached stuff will be maintained even when (custom)dataparallel replicate models or destroy them..? Are we on the same page now?

**Comment by dbolya** [comment]
Yeah and the reason they're maintained is because the original copy data parallel replicates from is now initialized.

**Comment by emjay73** [comment]
OK. Now I get it!
Thank you for everything!
